### ISSUE TITLE
Port twisted.python.filepath to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5903]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5903|
|Type|enhancement|
|Created|2012-08-22 19:40:20Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-python3-5903-4|

filepath should run on Python 3.3.

<details><summary>Searchable metadata</summary>

```
trac-id__5903 5903
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_filepath_python3_5903_4 branches/filepath-python3-5903-4
branch_author__itamarst__exarkun itamarst, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1345664420000000 1345664420000000
changetime__1346099700000000 1346099700000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35340]) Branching to 'filepath-python3-5903'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

This also includes a port of zippath. It depends on `twisted.python.util` and `runtime` which haven't yet been ported, but work sufficiently well to at least help the bootstrapping along.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filepath-python3-5903 is running now.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Sorry, not `twisted.python.util`, but rather `twisted.python.win32`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.

  1. The new comment in `admin/_twistedpython3.py` seems misleading and incorrect now.
  1. in `test_compat.py`, `assertNativeString` is missing a docstring
  1. There's no test coverage for the non-ASCII cases of `nativeString`.  Perhaps there should be?
  1. in `test_paths.py`, `test_zipArchiveRepr` seems like a very boring test.  Perhaps the test should actually literally encode the expected result?  It's clear that %r will equal %r, but that doesn't tell me it equals something useful.
  1. `filepath.py
    1. Rather than copying the implementation of `FancyEqMixin`, it would be better to put it into a new module and share it between util.py and filepath.py via import.
    1. zope import belongs between stdlib and twisted imports
    1. `_secureEnoughString` is used to generate path components, so it is safe for it to always return bytes, isn't it?
    1. int(most things) might be almost as redundant as long(most things).  The stat result seems to have all int attributes except for `st_dev` which is, for some reason, a long.  Do we even care?  Perhaps letting whatever the underlying type is pass through is sufficient.
  1. On a very general note, as long as we're burning all this time looking at every single line of Twisted, perhaps we should be taking note of specific improvements that would be desirable.  There are a lot of terrible things I see in this branch, but they're existing terrible things that don't get directly in the way of porting.  Still, it would be good to track that they exist, and a lot of them might make good introductory (aka "easy") tickets for new contributors.  It wouldn't surprise me if we came away from a ticket like this one with 5 - 10 other new tickets for tiny things (for example, not using `eval` in `test_permissionsShorthand`).
  1. I'm confused by the addition of `twisted/python/test/test_urlpath.py`.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [35359]) Address review comments 1-3. Refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

(In [35360]) Address review comment 5. Refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. I'm not sure what comment you mean exactly. I tried to improve it.
  2. Done.
  3. Done.
  4. Dropped - zippath is too difficult to do in this branch given zipfile only does Unicode on Python 3, so I am only porting filepath.
  5. Done.
  6. I opened [#5908](https://github.com/twisted/twisted/issues/5908), but don't want to spend the time looking for more silliness.
  7. urlpath, like zippath, used to be tested `test_paths`. Both of them are now tested in separate modules to ease testing.
  8. FilePath is now bytes-only on Python 3.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

  1. `twisted.python.compat.unicode` should be documented, I think.
  1. There's some conflicts in `admin/_twistedpython3.py`
  1. `twisted/python/_utilpy3.py` should be added to `_twistedpython3.py` right?
  1. `_stub_urandom` documents its return type as `str`.  `bytes` instead?
  1. I wonder what case `_stub_armor` is for.  When does `base64.urlsafe_b64encode` not exist?  Also, does `_stub_armor` work?  It just does `s.encode("hex")`, which will only work on unicode in Python 3.3.  It gets passed the output of `sha1.digest`, which seems to be bytes.  I guess there's no test coverage for the stub case of `_randomEnoughString`?
  1. I assume `test_zippath` and `test_urlpath` are just moved code with no other changes?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

(In [35378]) Branching to 'filepath-python3-5903-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

(In [35379]) Branching to 'filepath-python3-5903-3'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst set owner to @exarkun|
|-|-|

Re 1, it's not clear how I can add docstring without subclassing or defining of a new function, both of which would break uses like `isinstance`, so I just added a comment.

I dropped the two stub functions that were only necessary for versions of Python older than 2.4.

`_twistedpython3.py` had both issues fixed.

And, yes, those two test modules are just moved code, nothing has been changed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @itamarst|
|-|-|

> Re 1, it's not clear how I can add docstring without subclassing or defining of a new function, both of which would break uses like isinstance, so I just added a comment.

epytext has module docstring "@var" support, that's a good place for it.

> I dropped the two stub functions that were only necessary for versions of Python older than 2.4.
> _twistedpython3.py had both issues fixed.
> And, yes, those two test modules are just moved code, nothing has been changed.

Excellent.

Unfortunately I noticed another problem:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/filepath-python3-5903-3

Even ignoring the expected red (and purple today, hooray), that doesn't look good.  A Python 2.6 incompatibility?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst commented|
|-|-|

(In [35384]) Add @var, refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [35387]) Use SynchronousTestCase on Python 2.x

refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [35397]) Try resolving symlinks in the base part of the temporary path name, to avoid being confused on systems where /tmp is a symlink.

refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [35398]) Nope, that doesn't avoid the confusion.  Try using the current directory instead.

refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [35400]) Fix import problems

refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [35401]) Branching to 'filepath-python3-5903-4'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

Finally got a good buildbot run.  Fixed the Python 2.6 incompatibility mentioned above, as well as one OS X incompatibility related to temp file allocation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [35403]) Put that @var back in, how did it get dropped?

refs [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set status to `closed`|
|-|-|

(In [35404]) Merge filepath-python3-5903-4

Author: itamar, exarkun
Reviewer: exarkun
Fixes: [#5903](https://github.com/twisted/twisted/issues/5903)


Port `twisted.python.filepath` and its minor dependencies to Python 3.x.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set status to `reopened`|
|-|-|

(In [35408]) Revert r35404 - Python 2.6 test suite regressions

A bunch of failures like these:

exceptions.AttributeError: FilePathTestCase object has no attribute addCleanup

Reopens: [#5903](https://github.com/twisted/twisted/issues/5903)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_24"></a>|@itamarst set status to `closed`|
|-|-|

(In [35410]) Merge filepath-python3-5903-3

Author: itamar, exarkun
Reviewer: exarkun
Fixes: [#5903](https://github.com/twisted/twisted/issues/5903)


Port  and its minor dependencies to Python 3.x.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_25"></a>|@itamarst commented|
|-|-|

Version 4 of the branch apparently dropped a bunch of commits, and version 3 merged OK, and had all commits. I therefore merged version 3.
